### PR TITLE
fix overflow when training mDeberta in fp16

### LINKED ISSUE #77: mDeBERTa on HuggingFace hub does not seem to work
I really like the DeBERTa-v3 models and the monolingual models work very well for me. Weirdly enough, the multilingual model uploaded on the huggingface hub does not seem to work. I have a code for training multilingual models on XNLI, and the training normally works well (e.g. no issue with microsoft/Multilingual-MiniLM-L12-H384), but when I apply the **exact same code** to [mDeBERTa](https://huggingface.co/microsoft/mdeberta-v3-base), the model does not seem to learn anything. I don't get an error message, but the training results look like this:
Screenshot 2021-12-04 at 10 43 58

I've double checked by running the *exact same code* on multilingual-minilm and the training works, which makes me think that it's not an issue in the code (wrongly formatting the input data or something like that), but something went wrong when uploading mDeBERTa to the huggingface hub? Accuracy of exactly random 0.3333, 0 training loss at epoch 2 and NaN validation loss maybe indicates that the data is running through the model, but some parameters are not updating or something like that?

My environment is google colab; Transformers==4.12.5

### PR DESCRIPTION
**Pull request by sjrl**
# What does this PR do?

Fixes https://github.com/microsoft/DeBERTa/issues/77 (issue about transformers opened in Microsoft repo)

- This issue was originally raised in the https://github.com/microsoft/DeBERTa repo which had to do with mDeberta not being able to be trained using fp16. A fix for this was implemented in the Microsoft repo by @BigBird01 but did not yet make it to HuggingFace. I was interested in training mDeberta models on small hardware (e.g. a 3070, T4) so I updated the HF implementation with the changes from the Microsoft repo. I tried to only bring over the minimal changes needed to get the fp16 training to work. 

- I checked that existing tests passed and also used this code to successfully train an mDeberta model in fp16 on Squad2 that can be found [here](https://huggingface.co/sjrhuschlee/mdeberta-v3-base-squad2) which is not currently possible with the main branch of transformers. I'm unsure if there is a good way to add an additional test to make sure mDeberta-V3 training works in fp16 in the CI.

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Hey, based on the recommendations from the PR template (and git blame) I decided to tag @ArthurZucker and @sgugger in case you may be interested.

### CODE REVIEW

**Comment by amyeroberts** [comment]
cc @younesbelkada @ArthurZucker

**Review by ArthurZucker — COMMENTED** [review]
Hey! Thanks a lot for opening PR, and fixing a very old issue! 
We might have problems with this with quantization, as the errors are probably gonna be different. Could you check if using `torch.float16` or `load_in_8bits` gives the same outputs. (this might fix training but maybe break inference? let's just check that everything runs correctly!)

See #22444

**Comment by sjrl** [comment]
I used this code block to check results
```python
import torch
from transformers import AutoTokenizer, AutoModelForQuestionAnswering
from transformers.pipelines import QuestionAnsweringPipeline

tokenizer = AutoTokenizer.from_pretrained("sjrhuschlee/mdeberta-v3-base-squad2")
model = AutoModelForQuestionAnswering.from_pretrained(
    "sjrhuschlee/mdeberta-v3-base-squad2",
#     torch_dtype=torch.float16,
#     torch_dtype=torch.bfloat16,
#     load_in_8bit=True,
)
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))  # device=... was removed for 8bit
```

**Running on Main Branch**

Running the above code using `torch.float16` on the main branch gives me no answer
```python
# with torch.float16
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# []
```
Running with `torch.bfloat16` and `torch.float32` gives me the expected answer
```python
# with torch.bfloat16
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# {'score': 0.98369300365448, 'start': 33, 'end': 41, 'answer': ' Berlin.'}

# with torch.float32
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# {'score': 0.9850791096687317, 'start': 33, 'end': 41, 'answer': ' Berlin.'}
```
Also running in `8bit` works
```python
# with load_in_8bit=True
pipe = QuestionAnsweringPipeline(model, tokenizer)
# {'score': 0.9868391752243042, 'start': 33, 'end': 41, 'answer': ' Berlin.'}
```

**Running on the PR**
The change in this PR also enables mDeberta models to run at inference in `torch.float16` which wasn't possible before. And it doesn't look to affect any of the other dtypes.
```python
# with torch.float16
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# {'score': 0.9848804473876953, 'start': 33, 'end': 41, 'answer': ' Berlin.'}

# with torch.bfloat16
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# {'score': 0.9841369986534119, 'start': 33, 'end': 41, 'answer': ' Berlin.'}

# with torch.float32
pipe = QuestionAnsweringPipeline(model, tokenizer, device=torch.device("cuda:0"))
# {'score': 0.9850791096687317, 'start': 33, 'end': 41, 'answer': ' Berlin.'}

# with load_in_8bit=True
pipe = QuestionAnsweringPipeline(model, tokenizer)
# {'score': 0.9870386719703674, 'start': 33, 'end': 41, 'answer': ' Berlin.'}
```

**Review by ArthurZucker — APPROVED** [review]
Cool! seems like a very subtle but effective change! Pinging @amyeroberts for a second pair of eyes

**Comment by sjrl** [comment]
I also noticed that the TF implementation in DebertaV2 has the same line https://github.com/huggingface/transformers/blob/2e2088f24b60d8817c74c32a0ac6bb1c5d39544d/src/transformers/models/deberta_v2/modeling_tf_deberta_v2.py#L678-L679
I'm not too familiar with TF though so I'm not sure if this change should be made there as well.

**Comment by Rocketknight1** [comment]
Yes, we support mixed-precision float16/bfloat16 training in TensorFlow, but in general we still expect a 'master' copy of the weights to remain in float32. We're planning some exploration to see if we can get Keras to accept full (b)float16 training, but it might require some refactoring!

**Review by younesbelkada — APPROVED** [review]
LGTM

**Comment by sjrl** [comment]
@Rocketknight1 should I go ahead and update the TF implementation as well then?

**Comment by Rocketknight1** [comment]
@sjrl Yes please! Better numerical stability will be nice to have once we've enabled full float16 training

**Review by amyeroberts — APPROVED** [review]
Thanks for this contribution and making our models more stable ❤️

**Comment by amyeroberts** [comment]
@sjrl  - Are there any other changes to add? Otherwise I think we're good to merge :)
